src/pint/models/model_builder.py

```diff
@@ -2,317 +2,366 @@
-
-    def __str__(self):
-        result = "Model name : " + self.name + "\n"
-        result += "Components in the model : \n"
-        for c in self.select_comp:
-            result += "    " + str(c) + "\n"
-        if self.model_instance is not None:
-            result += "Read parameters from : " + self.parfile + "\n"
-            result += "The model instance is :\n" + str(self.model_instance)
-
-        return result
-
-    def preprocess_parfile(self, parfile):
-        """Preprocess the par file.
+    def __init__(self):
+        super().__init__()
+        # Validate the components
+        self._validate_components()
+        self.default_components = ["SolarSystemShapiro"]
+
+    def __call__(self, parfile):
+        param_inpar, repeat_par = self.parse_parfile(parfile)
+        selected, conflict, unrec_param = self.choose_model(param_inpar)
+        selected.update(set(self.default_components))
+        # Report conflict
+        if len(conflict) != 0:
+            self._report_conflict(conflict)
+        # Make timing model
+        cps = [self.components[c] for c in selected]
+        tm = TimingModel(components=cps)
+        # Add indexed parameters
+        leftover_params = set(unrec_param)
+        # Give repeatable parameters an index.
+        for k, v in repeat_par.items():
+            for ii in range(len(v)):
+                leftover_params.add(k + str(ii + 1))
+        tm, unknown_param = self._add_indexed_params(tm, leftover_params)
+        return tm
+
+    def _validate_components(self):
+        """Validate the built-in component.
+
+        This function validates if there is a subset parameter conflict in the
+        Components. Normally, one component's parameter should not be a subset
+        of another component's parameter list. Otherwise, the model builder does
+        not have an unique choice of component. Currently, Pulsar binary does
+        not follow the same rule. They are specified by the `BINARY` parameter.
+        """
+        for k, v in self.components.items():
+            superset = self._is_subset_component(v)
+            if superset is not None:
+                m = (
+                    f"Component {k}'s parameter is a subset of component"
+                    f" {superset}. Module builder will have trouble to "
+                    f" select the component. If component {k} is a base"
+                    f" class, please set register to 'False' in the class"
+                    f" of component {k}."
+                )
+                if v.category == "pulsar_system":
+                    # The pulsar system will be selected by parameter BINARY
+                    continue
+                else:
+                    raise ComponentConflict(m)
 
+    def _get_component_param_overlap(self, component):
+        """Check if one component's parameters are overlaped with another
+        component.
+
+        Parameters
+        ----------
+        components: component object
```

Comment:
Please use the actual fully qualified class name so Sphinx can make a link. Also this is not the name of the argument in the function signature. (Parameter names are part of a function's public API in python.)